symengine/subs.h

```diff
@@ -276,6 +276,229 @@ class SubsVisitor : public BaseVisitor<SubsVisitor, XReplaceVisitor>
+                }
+            } else {
+                fast_exec = true;
+                RCP<const Basic> exp, t;
+                Mul::as_base_exp(factor, outArg(exp), outArg(t));
+                Mul::dict_add_term_new(outArg(coef), d, exp, t);
+            }
+        }
+        if (fast_exec) {
+            result_ = Mul::from_dict(coef, std::move(d));
+            return;
+        }
+        for (const auto &iter : subs_dict_) {
+            d.clear();
+            bool exists = true;
+            auto sub1 = iter.first;
+            auto rep = iter.second;
+            if (is_a<Mul>(*sub1)) {
+                RCP<const Mul> subst = rcp_static_cast<const Mul>(sub1);
+                for (auto &p : subst->get_dict()) {
+                    auto it = dict.find(p.first);
+                    RCP<const Basic> diff_;
+                    if (it != dict.end())
+                        diff_ = sub(it->second, p.second);
+                    if (it == dict.end()
+                        || down_cast<const Number &>(*diff_).is_negative()) {
+                        exists = false;
+                        break;
+                    } else {
+                        if (!down_cast<const Number &>(*diff_).is_zero())
+                            Mul::dict_add_term_new(outArg(coef), d,
+                                                   sub(it->second, p.second),
+                                                   p.first);
+                    }
+                }
+                if (exists) {
+                    for (const auto &p : dict) {
+                        auto it = subst->get_dict().find(p.first);
+                        if (it == subst->get_dict().end())
+                            Mul::dict_add_term_new(outArg(coef), d, p.second,
+                                                   p.first);
+                    }
+                    if (is_a_Number(*rep)) {
+                        if (down_cast<const Number &>(*rep).is_zero()) {
+                            result_ = rep;
+                            return;
+                        }
+                        imulnum(outArg(coef),
+                                rcp_static_cast<const Number>(rep));
+                    } else if (is_a<Mul>(*rep)) {
+                        RCP<const Mul> tmp = rcp_static_cast<const Mul>(rep);
+                        imulnum(outArg(coef), tmp->get_coef());
+                        for (const auto &q : tmp->get_dict()) {
+                            Mul::dict_add_term_new(outArg(coef), d, q.second,
+                                                   q.first);
+                        }
+                    } else {
+                        RCP<const Basic> exp, t;
+                        Mul::as_base_exp(rep, outArg(exp), outArg(t));
+                        Mul::dict_add_term_new(outArg(coef), d, exp, t);
+                    }
+                } else
+                    d = x.get_dict();
+            } else if (is_a<Pow>(*sub1)) {
```

Comment:
This is handled in the first part? Why check again?

Reply:
The first part is the earlier implementation that only looks for exact matches, and not partial matches which should be handled using `subs`.
That part is retained only for the fast execution speeds in case of exact matches as it doesn't involve many heavy computations.